File: apps/web/types/certification.ts
/**
 * Evidence validation and certification types
 */

export interface EvidenceBundle {
  model_name: string
  model_version: string
  data_governance?: {
    data_sources: string[]
    data_lineage_documented: boolean
    consent_obtained: boolean
    retention_policy_days: number
  }
  explainability?: {
    feature_importance_available: boolean
    model_cards_documented: boolean
    interpretability_methods: string[]
  }
  robustness?: {
    adversarial_testing_performed: boolean
    bias_audit_score: number // 0-100
    performance_on_minority_groups: number // 0-100
  }
  operational?: {
    monitoring_enabled: boolean
    incident_response_plan: boolean
    model_degradation_threshold: number
  }
  alignment?: {
    privacy_controls: boolean
    fairness_commitments: boolean
    human_oversight: boolean
  }
}

export interface ValidationError {
  field: string
  message: string
  severity: 'error' | 'warning'
}

export interface ValidationResult {
  valid: boolean
  errors: ValidationError[]
  warnings: ValidationError[]
  timestamp: string
}

export interface CertificationScore {
  alignment_score: number // 0-100
  robustness_score: number // 0-100
  data_governance_score: number // 0-100
  explainability_score: number // 0-100
  operational_risk_score: number // 0-100
  overall_score: number // 0-100
  certification_level: 'bronze' | 'silver' | 'gold' | 'platinum' | 'failed'
  recommendations: string[]
}

export interface Attestation {
  id: string
  model_name: string
  model_version: string
  organization: string
  issued_date: string
  valid_until: string
  scores: CertificationScore
  validation_results: ValidationResult
  certificate_hash: string
  verification_url: string
}

export interface AttestationVerification {
  id: string
  valid: boolean
  model_name: string
  model_version: string
  organization: string
  issued_date: string
  certification_level: string
  score: number
  verified_timestamp: string
}

File: apps/api/app/models/certification.py
"""
Certification and evidence validation models.
"""
from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, Field
from enum import Enum

class CertificationLevel(str, Enum):
    """Certification levels."""
    FAILED = "failed"
    BRONZE = "bronze"
    SILVER = "silver"
    GOLD = "gold"
    PLATINUM = "platinum"

class EvidenceBundleData(BaseModel):
    """Evidence bundle data model - matches frontend types."""
    model_name: str
    model_version: str
    data_governance: Optional[dict] = None
    explainability: Optional[dict] = None
    robustness: Optional[dict] = None
    operational: Optional[dict] = None
    alignment: Optional[dict] = None

class ValidationError(BaseModel):
    """Validation error."""
    field: str
    message: str
    severity: str  # "error" or "warning"

class ValidationResult(BaseModel):
    """Result of evidence validation."""
    valid: bool
    errors: List[ValidationError] = []
    warnings: List[ValidationError] = []
    timestamp: str

class CertificationScore(BaseModel):
    """Certification scores for a model."""
    alignment_score: float = Field(ge=0, le=100)
    robustness_score: float = Field(ge=0, le=100)
    data_governance_score: float = Field(ge=0, le=100)
    explainability_score: float = Field(ge=0, le=100)
    operational_risk_score: float = Field(ge=0, le=100)
    overall_score: float = Field(ge=0, le=100)
    certification_level: CertificationLevel
    recommendations: List[str] = []

class Attestation(BaseModel):
    """Attestation certificate."""
    id: str
    model_name: str
    model_version: str
    organization: str
    issued_date: str
    valid_until: str
    scores: CertificationScore
    validation_results: ValidationResult
    certificate_hash: str
    verification_url: str

    class Config:
        from_attributes = True

class AttestationVerification(BaseModel):
    """Attestation verification response."""
    id: str
    valid: bool
    model_name: str
    model_version: str
    organization: str
    issued_date: str
    certification_level: str
    score: float
    verified_timestamp: str

class ValidateRequest(BaseModel):
    """Validate evidence request."""
    evidence: EvidenceBundleData

class CertificationRequest(BaseModel):
    """Request certification scores."""
    evidence: EvidenceBundleData
    validation_result: ValidationResult

class AttestationRequest(BaseModel):
    """Request attestation issuance."""
    evidence: EvidenceBundleData
    scores: CertificationScore
    organization_id: str
